### ISSUE TITLE
incorrect comparison in XMG logic

### ISSUE THREAD

**Issue by mfernan2** [lsils/mockturtle#256]
When compiling with `-Wparentheses` GCC 9.2.0 has this to say:

```sh
In file included from /home/matthew/project/mockturtle/include/mockturtle/algorithms/node_resynthesis/xmg_npn.hpp:47:0,
                 from /home/matthew/project/mockturtle/include/mockturtle/mockturtle.hpp:71,
                 from /home/matthew/project/project/main.cc:8:
/home/matthew/project/mockturtle/include/mockturtle/algorithms/node_resynthesis/../../networks/xmg.hpp: In member function ‘mockturtle::xmg_network::signal mockturtle::xmg_network::create_maj(mockturtle::xmg_network::signal, mockturtle::xmg_network::signal, mockturtle::xmg_network::signal)’:
/home/matthew/project/mockturtle/include/mockturtle/algorithms/node_resynthesis/../../networks/xmg.hpp:300:23: warning: suggest parentheses around comparison in operand of ‘==’ [-Wparentheses]
     else if ( a.index == b.index == c.index )
               ~~~~~~~~^~~~~~~~~~
```

This looks to me like a legitimate problem. Is this supposed to read `a.index == b.index && b.index == c.index`?

**Comment by hriener** [comment]
Thanks for having a look. In general, it would be great to fix all warnings in `mockturtle`. Any help in this context is very welcome!

I had a quick look at the problem. The suggested fix looks like unreachable code to me. If `a.index == b.index` and `b.index == c.index` then also the previous two if-conditions should be already satisfied and the method should have returned (and the same seems to be true for line 383).

**Comment by mfernan2** [comment]
Ah yes you're right, I did not actually look at the preceding branches. There's yet another case at [line 627](https://github.com/lsils/mockturtle/blob/master/include/mockturtle/networks/xmg.hpp#L627).

Looking at the surrounding code for all these three situations, I don't think this is branch is *wrong*, but simply redundant because the three things being compared are sorted at this point (a ≤ b ≤ c). Seems reasonable to delete the redundant code. I.e.

```diff
--- xmg.hpp     2020-03-09 08:59:33.000000000 -0700
+++ xmg.hpp    2020-03-09 08:59:02.000000000 -0700
@@ -297,10 +297,6 @@ public:
     {
       return ( b.complement == c.complement ) ? b : a;
     }
-    else if ( a.index == b.index == c.index )
-    {
-      return ( a.complement == b.complement ) ? a : c;
-    }

     /*  complemented edges minimization */
     auto node_complement = false;
@@ -380,10 +376,6 @@ public:
     {
       return a ^ fcompl;
     }
-    else if ( ( a.index == b.index ) && ( b.index == c.index ) )
-    {
-      return a ^ fcompl;
-    }

     storage::element_type::node_type node;
     node.children[0] = a;
@@ -624,10 +616,6 @@ public:
       {
         return std::make_pair( n, child0 ^ node_complement );
       }
-      else if ( ( child0.index == child1.index ) && ( child1.index == child2.index ) )
-      {
-        return std::make_pair( n, child0 ^ node_complement );
-      }
     }

     // node already in hash table
```

**Comment by hriener** [comment]
Could you provide a PR for the changes?

**Comment by mfernan2** [comment]
I would love to, but unfortunately I'm behind a rather draconian corporate policy firewall. Based on past estimates, getting clearance to submit a PR will take >6 months. I'm happy to go down the path of getting approvals for this if you prefer, but it will probably be some time before I'll be able to post the PR.

**Comment by mfernan2** [comment]
The second two instances of this were fixed in c2fd7c82 by @msoeken.

**Comment by hriener** [comment]
Fixed! Thanks!
